### ISSUE TITLE
Slow Training speed on Custom Dataset

### ISSUE THREAD

**Issue by souvik0306** [haleqiu/AirIMU#18]
Hi!  

I'm trying to implement **AirIMU** using the **University of Zurich FPV Drone Racing Dataset** ([link](https://fpv.ifi.uzh.ch/)). However, I'm experiencing extremely slow training speeds. Each epoch is taking approximately **18 minutes** on an **NVIDIA A6000 GPU**, which feels unusually long given the limited data I'm using.

<img width="1137" height="283" alt="Image" src="https://github.com/user-attachments/assets/27a8d83d-736c-4ae1-8d8d-40f746a7313b" />

### Setup
- **Max epochs:** 50  
- **GPU:** NVIDIA RTX A6000  
- **Training sequences:** 3 sequences using 500Hz IMU data 
  Each sequence is approximately **50–70 seconds** long  

### My Goal
1. Train the model using **AirIMU**  
2. Later use it with **Air-IO** for the **EKF module**

### Training Configuration

```yaml
train:
{
    mode: train
    calib: False
    data_list:
    [   
        {
            name: Uzurich
            window_size: 1000
            step_size: 3
            data_root: Uzurich
            data_drive: [
                indoor_forward_3_snapdragon_with_gt, 
                indoor_forward_7_snapdragon_with_gt,
                indoor_45_9_snapdragon_with_gt
            ]
        },
    ]
    gravity: 9.80511
}
```

My Questions

- Is this training time expected given the sequence length and step size?
- Am I doing something fundamentally wrong using my custom datasets script? [Please find it here - Link](https://github.com/souvik0306/AirIMU/blob/main/datasets/Uzurichdataset.py)
- Are there any recommended changes to improve training speed while maintaining accuracy?

Any help or pointers would be appreciated!

**Comment by haleqiu** [comment]
@souvik0306 [Quick Reply] Yes, it will run slower with a longer window size since the gradient needs to propagate through the entire IMU sequence of that length. If you’re also training with covariance, it will be slower still. You can try using a larger step size or increasing the batch size to improve speed.

By the way, you may not neccesaary to reach 50 epoches if it already converge.

**Comment by souvik0306** [comment]
> [@souvik0306](https://github.com/souvik0306) [Quick Reply] Yes, it will run slower with a longer window size since the gradient needs to propagate through the entire IMU sequence of that length. If you’re also training with covariance, it will be slower still. You can try using a larger step size or increasing the batch size to improve speed.
> 
> By the way, you may not neccesaary to reach 50 epoches if it already converge.

Thanks @haleqiu, let me incorporate the changes and let you know!

**Comment by souvik0306** [comment]
Thanks again for the earlier response. I was able to speed up training a bit by increasing the batch size and running a few sequences as a sanity check.

### Follow-up query

I'm now trying to compare the performance of:
- AirIMU (standalone)
- AirIO (Neural Net only)
- AirIO (NN + EKF)

on the EuRoC and Blackbird datasets. The idea is to check whether AirIMU alone is sufficient for IO tasks, or whether the full AirIO pipeline gives a significant advantage.

### Setup

I used the pretrained models provided here:  
https://github.com/Air-IO/Air-IO?tab=readme-ov-file#get-pre-trained-models--results

For evaluating AirIMU standalone on EuRoC, I used the same dataset sequences as in the original code and set `usegtrot` to `True`.

```bash
python evaluation/evaluate_state.py \
--dataconf configs/datasets/BaselineEuroc/Euroc_1000.conf \
--exp experiments/EuRoC/codenet/ \
--usegtrot
```

### Results (AirIMU only)
Here is the result on MH_02_easy (Euroc) :

```
    {
        "name": "MH_02_easy",
        "use_gt_rot": true,
        "AOE(raw)": 114.78726556143508,
        "ATE(raw)": 349.18007169558456,
        "AVE(raw)": 6.3011035323549125,
        "ROE(raw)": 4.579851485915311,
        "RTE(raw)": 0.08258470292131824,
        "RVE(raw)": 0.16357963874607198,
        "RP_RMSE(raw)": 0.083225659271418,
        "RV_RMSE(raw)": 0.1647314016643741,
        "RO_RMSE(raw)": 4.580001407359935,
        "O_RMSE(raw)": 123.0527889681259,
        "AOE(AirIMU)": 3.937127270719086,
        "ATE(AirIMU)": 52.88760510156253,
        "AVE(AirIMU)": 0.9865360323565787,
        "ROE(AirIMU)": 0.06805566583013589,
        "RTE(AirIMU)": 0.020856224351746264,
        "RVE(AirIMU)": 0.03612695936383537,
        "RP_RMSE(AirIMU)": 0.023095063108190494,
        "RV_RMSE(AirIMU)": 0.039931005347747524,
        "RO_RMSE(AirIMU)": 0.07592708398156894,
        "O_RMSE(AirIMU)": 4.26045266869652
    }
```

###  Issue (@haleqiu)

The ATE from AirIMU seems much higher than expected based on the metrics in the AirIMU paper. By contrast, AirIO NN and NN+EKF give lower ATE and train faster. I’m trying to understand:

1. What factors could be contributing to the higher ATE from AirIMU?
2. Is there a recommended evaluation pipeline for ATE to match the reported numbers from the AirIMU paper?
3. Do we need to fine-tune further, even when using the provided pretrained weights?

Would appreciate any guidance or suggestions.

Thanks again for your great work!

**Comment by haleqiu** [comment]
> By contrast, AirIO NN and NN+EKF give lower ATE and train faster.

Of course, AirIO is expected to yield a lower ATE since it is a full odometry system.

AirIMU, on the other hand, focuses on preintegration, where ATE is not a very meaningful metric. Due to the inherent drift from double integration, ATE will inevitably grow. I would suggest not relying on ATE as a reference in this case.

**Comment by souvik0306** [comment]
> > By contrast, AirIO NN and NN+EKF give lower ATE and train faster.
> 
> Of course that AirIO should give a lower ATE as it is a full odometry.
> 
> AirIMU focus on the preintegration, the ATE is not very useful here. Integration will inevitably drift due to the nature of double integration. I will suggest you not taking the ATE as a reference.

I was wondering how we can reach such metrics for AirIMU? Since this Abalation study focuses on the ATE and the metrics looks quite promising - 
<img width="963" height="390" alt="Image" src="https://github.com/user-attachments/assets/cfe827c0-c03b-4c24-b3b9-8f84c0edc0a0" />

**Comment by haleqiu** [comment]
@souvik0306 Hi, That is the Pose graph optimization result which is fusing with other sensors. It is not the IMU only result. 
You can fuse it with visual odometry or any other sensors.

**Comment by souvik0306** [comment]
> [@souvik0306](https://github.com/souvik0306) Hi, That is the Pose graph optimization result which is fusing with other sensors. It is not the IMU only result. You can fuse it with visual odometry or any other sensors.

Oh I see, thanks a lot!

**Comment by haleqiu** [comment]
Close this as inactive
